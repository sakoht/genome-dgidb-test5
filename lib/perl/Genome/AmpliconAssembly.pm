package Genome::AmpliconAssembly;

use strict;
use warnings;

use Genome;

use Carp 'confess';
use Cwd 'abs_path';
use Data::Dumper 'Dumper';
require File::Copy;
require Genome::Model::Tools::Consed::Directory;
require Genome::AmpliconAssembly::Amplicon;

my %ATTRIBUTES = (
    directory => {
        is => 'Text',
        is_optional => 1,
        doc => 'Base directory for the amplicon assembly.',
    },
    description => {
        is => 'Text',
        is_optional => 1,
        doc => 'A brief description of the amplicon assembly. Helps identify in messaging. Default is the directory.',
    },
    sequencing_center => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_center,
        doc => 'Sequencing Center that the amplicons were sequenced. Currently supported centers: '.join(', ', __PACKAGE__->valid_sequencing_centers).' (default: '.__PACKAGE__->default_sequencing_center.').',
    },
    sequencing_platform => {
        is => 'Text',
        is_optional => 1,
        default_value => __PACKAGE__->default_sequencing_platform,
        doc => 'Platform upon whence the amplicons were sequenced. Currently supported platforms: '.join(', ', __PACKAGE__->valid_sequencing_platforms).' (default: '.__PACKAGE__->default_sequencing_platform.').',
    },
    assembly_size => { # amplicon_size
        is => 'Integer',
        is_optional => 1,
        default_value => 1364,
        doc => 'Expected assembly size for an assembled amplicon.',
    },
    subject_name => {
        is => 'Text',
        is_optional => 1,
        doc => 'The subject name of the underlying data. Used primarily in naming files (etc) that represent the entire amplicon assembly, like an assembled fasta of each amplicon.',
    },
    exclude_contaminated_amplicons => {
        is => 'Boolean',
        is_optional => 1,
        default_value => 0,
        doc => 'When getting amplicons, exclude those that have a contaminated read(s). Only for "gsc" generated reads. Default is to include all amplicons.',
    },
    only_use_latest_iteration_of_reads => {
        is => 'Boolean',
        is_optional => 1,
        default_value => 0,
        doc => 'When getting reads for amplicons, only use the most recent iteration for each primer. Only for "gsc" generated reads. Default is to include all reads for each amplicon.',
    },
);

#< UR >#
class Genome::AmpliconAssembly {
    is => 'UR::Object',
    id_by => 'directory',
    has => [
    %ATTRIBUTES,
    _additional_properties => {
        is => 'HASH',
        default_value => {},
        doc => 'Properties added from operations performed on an amplicon assembly.',
    },
    ],
};

sub get { 
    my ($class, %params) = @_;

    # Do not support getting with any other params...
    my $directory = delete $params{directory};
    if ( %params ) { 
        $class->error_message('Can\'t get with additional parameters, only directory.');
        return;
    }
    
    # Validate directory
    unless ( Genome::Sys->validate_existing_directory($directory) ) {
        $class->error_message("Can't validate amplicon assembly directory. See above error.");
        return;
    }

    # UR Cache
    my ($self) = $class->SUPER::get(directory => $directory);
    return $self if $self;
    
    # Properties file
    return $class->_create_from_saved_properties($directory);
}

sub create {
    my ($class, %params) = @_;

    $params{directory} = Cwd::abs_path($params{directory});
    unless ( Genome::Sys->validate_existing_directory($params{directory}) ) {
        $class->error_message("Can't validate amplicon assembly directory. See above error.");
        return;
    }

    # Get UR cache/properties file
    #my ($self) = $class->get(directory => $params{directory});
    #return $self if $self;

    my $self = $class->SUPER::create(%params)
        or return;
    
    unless ( $self->description ) {
        $self->description( $self->directory );
    }

    $self->_validate_properties
        or return;

    $self->create_directory_structure
        or return;
    
    $self->_save_properties
        or return;

    return $self;
}

#< Properties >#
sub _properties_file {
    my ($self, $directory) = @_;

    if ( not $directory and ref $self ) {
        $directory = $self->directory;
    }

    confess "No directory given to get properties file." unless $directory;

    return $directory.'/properties.stor';
}

sub _create_from_saved_properties {
    my ($class, $directory) = @_;

    my $properties_file = $class->_properties_file($directory);
    return unless -e $properties_file;
    
    my $properties = Storable::retrieve($properties_file);
    unless ( $properties ) {
        confess "No properties in amplicon assembly properties file.";
    }

    return $class->SUPER::create(%$properties)
}

sub _validate_properties {
    my $self = shift;

    for my $attribute (qw/ sequencing_platform sequencing_center /) {
        unless ( $self->validate_attribute_value($attribute) ) {
            $self->delete;
            return;
        }
    }

    unless ( $self->sequencing_center eq 'gsc' ) {
        if ( my @attrs = grep { $self->$_ } gsc_only_attributes() ) {
            $self->error_message(
                sprintf(
                    'Indicated unsupported attributes (%s) for sequencing center (%s).',
                    join(', ', @attrs),
                    $self->sequencing_center,
                )
            );
            $self->delete;
            return;
        }
    }

    return 1;
}

sub _save_properties {
    my $self = shift;

    my %properties;
    for my $attr ( $self->attribute_names ) {
        $properties{$attr} = $self->$attr;
    }

    $properties{_additional_properties} = $self->_additional_properties;

    my $properties_file = $self->_properties_file;
    unlink $properties_file if -e $properties_file;
    eval {
        Storable::store(\%properties, $properties_file);
    };
    
    if ( $@ ) {
        confess "Can't save properties to file ($properties_file): $@";
    }
    
    return 1;
}

sub update_additional_properties {
    my ($self, %properties) = @_;

    confess "No properties given to update additional properties." unless %properties;

    for my $property ( keys %properties ) {
        $self->_additional_properties->{$property} = $properties{$property};
    }
    
    return $self->_save_properties;
}

#< Atributes >#
sub attributes {
    return %ATTRIBUTES;
}

sub attribute_names {
    return keys %ATTRIBUTES;
}

sub attributes_without_default_values {
    my $class = shift;

    my %attributes;
    for my $name ( attribute_names() ) {
        $attributes{$name} = $class->get_attribute_for_name($name);
        delete $attributes{$name}->{default_value};
    }

    return %attributes;
}

sub get_attribute_for_name {
    my ($class, $name) = @_;

    confess "No attribute name given to get attribute" unless $name;
    confess "No attribute found for name ($name)" unless exists $ATTRIBUTES{$name};

    return $ATTRIBUTES{$name};
}

sub helpful_methods {
    return (qw/ 
        chromat_dir phd_dir edit_dir
        consed_directory create_directory_structure
        get_amplicons 
        amplicon_fasta_types amplicon_bioseq_method_for_type
        fasta_file_for_type qual_file_for_type
        assembly_fasta reads_fasta processed_fasta 
        /);
}

sub validate_attribute_value {
    my ($self, $attribute) = @_;

    my $value = $self->$attribute;
    unless ( defined $value ) { # all attrs have defaults, make sure it is set
        $self->error_message("No value for $attribute given");
        return;
    }
    
    my $valid_values_method = 'valid_'.$attribute.'s';
    unless ( grep { $value eq $_ } $self->$valid_values_method ) {
        $self->error_message(
            sprintf(
                'Invalid %s %s. Valid %ss: %s',
                $attribute,
                $value,
                $attribute,
                join(', ',valid_sequencing_centers()),
            )
        );
        return;
    }

    return 1;
}

# Sequencing Centers
sub valid_sequencing_centers {
    return (qw/ gsc broad /);
}

sub default_sequencing_center {
    return (valid_sequencing_centers)[0];
}

# Sequencing Platforms
sub valid_sequencing_platforms {
    return (qw/ sanger /);
}

sub default_sequencing_platform {
    return (valid_sequencing_platforms)[0];
}

# GSC Only Attributes
sub gsc_only_attributes {
    return (qw/ only_use_latest_iteration_of_reads exclude_contaminated_amplicons /);
}

#< DEPRACATED FIXME #>
sub assembler { return 'phredphrap' };

#< DIRS >#
sub consed_directory {
    my $self = shift;

    unless ( $self->{_consed_directory} ) {
        $self->{_consed_directory} = Genome::Model::Tools::Consed::Directory->create(directory => $self->directory);
    }

    return $self->{_consed_directory};
}

sub create_directory_structure {
    my $self = shift;

    $self->consed_directory->create_extended_directory_structure
        or return;

    return 1;
}

sub edit_dir {
    return $_[0]->consed_directory->edit_dir;
}
    
sub phd_dir {
    return $_[0]->consed_directory->phd_dir;
}
    
sub chromat_dir {
    return $_[0]->consed_directory->chromat_dir;
}

sub fasta_dir {
    return $_[0]->consed_directory->fasta_dir;
}

#< FASTA >#
my %_fastas_and_amplicon_bioseq_methods = (
    reads => 'get_bioseqs_for_raw_reads',
    processed => 'get_bioseqs_for_processed_reads',
    assembly => 'get_assembly_bioseq',
    oriented => 'get_oriented_bioseq',
);

sub amplicon_fasta_types {
    return keys %_fastas_and_amplicon_bioseq_methods;
}

sub amplicon_bioseq_method_for_type {
    return $_fastas_and_amplicon_bioseq_methods{$_[1]};
}

sub fasta_file_for_type {
    my ($self, $type) = @_;

    return sprintf(
        '%s/%s%s.fasta',
        $self->fasta_dir,
        ( defined $self->subject_name ? $self->subject_name.'.' : '' ),
        $type,
    );
}

sub qual_file_for_type {
    return $_[0]->fasta_file_for_type($_[1]).'.qual';
}

#< Amplicons >#
sub get_amplicons {
    my $self = shift;

    my $method = sprintf(
        '_get_amplicons_and_read_names_for_%s_%s', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );

    my $iterator = $self->$method;
    
    my $amplicons = $self->$method;
    unless ( $amplicons and %$amplicons ) {
        $self->error_message(
            sprintf('No amplicons found in chromat_dir of directory (%s)', $self->directory) 
        );
        return;
    }

    # Only use most recent read
    if ( $self->only_use_latest_iteration_of_reads ) {
        $self->_remove_old_read_iterations_from_amplicons($amplicons)
            or return;
    }
    
    if ( $self->exclude_contaminated_amplicons ) {
        $self->_remove_contaminated_amplicons($amplicons)
            or return;
    }
    
    my @amplicons;
    my $edit_dir = $self->edit_dir;
    for my $name ( sort { $a cmp $b } keys %$amplicons ) {
        push @amplicons, Genome::AmpliconAssembly::Amplicon->create(
            name => $name,
            reads => $amplicons->{$name},
            directory => $edit_dir,
        );
    }

    return \@amplicons;
}

sub _get_amplicons_and_read_names_for_gsc_sanger {
    my $self = shift;

    my $read_iterator = $self->_get_read_name_iterator
        or return;

    my %amplicons;
    while ( my $read_name = $read_iterator->() ) {
        my $amplicon_name = $self->_get_amplicon_name_for_gsc_sanger_read_name($read_name)
            or next; # Assume conversion logic is good, and returns undef cuz it is not a read
        push @{$amplicons{$amplicon_name}}, $read_name;
    }

    return \%amplicons;
}

sub _get_amplicons_and_read_names_for_broad_sanger {
    my $self = shift;

    my $read_iterator = $self->_get_read_name_iterator
        or return;

    my %amplicons;
    while ( my $read_name = $read_iterator->() ) {
        my $amplicon = $self->_get_amplicon_name_for_broad_sanger_read_name($read_name)
            or return; # Assume conversion logic is good, and returns undef cuz it is not a read
        push @{$amplicons{$amplicon}}, $read_name;
    }

    return  \%amplicons;
}

sub _get_read_name_iterator {
    my $self = shift;

    my $dh = Genome::Sys->open_directory( $self->chromat_dir );
    unless ( $dh ) {
        $self->error_message("Can't open chromat dir to get reads. See above error.");
        return;
    }

    return sub{
        while ( my $name = $dh->read ) {
            next if $name =~ m#^\.#;
            $name =~ s#\.gz##;
            return $name;
        }
        $dh->close;
        return;
    }
}

sub _remove_old_read_iterations_from_amplicons {
    my ($self, $amplicons) = @_;

    for my $amplicon_name ( keys %$amplicons ) {
        my %reads;
        for my $read_name ( @{$amplicons->{$amplicon_name}} ) {
            my $read = $self->_get_gsc_sequence_read($read_name);
            confess "Can't get read for name ($read_name). This is required to remove old read iterations from amplicons" unless $read;
            
            my $read_id = $amplicon_name.$read->primer_code;
            if ( exists $reads{$read_id} ) {
                my $date_compare = UR::Time->compare_dates(
                    '00:00:00 '.$read->run_date,
                    '00:00:00 '.$reads{$read_id}->run_date,
                ); #returns -1, 0, or 1
                #print "RUN DATE $read_name => ".$read->run_date."($date_compare)\n";
                $reads{$read_id} = $read if $date_compare eq 1;
            }
            else {
                $reads{$read_id} = $read;
            }
        }
        $amplicons->{$amplicon_name} = [
        sort { $a cmp $b } 
        map { $_->trace_name }
        values %reads 
        ];
    }

    return 1;
}

sub _remove_contaminated_amplicons {
    my ($self, $amplicons) = @_;

    AMPLICON: for my $amplicon_name ( keys %$amplicons ) {
        for my $read_name ( @{$amplicons->{$amplicon_name}} ) {
            my $read = $self->_get_gsc_sequence_read($read_name);
            confess "Can't get read for name ($read_name). This is required to remove contaminated amplicons." unless $read;
            my $screen_reads_stat = $read->get_screen_read_stat_hmp;
            if ( $screen_reads_stat and $screen_reads_stat->is_contaminated ) {
                delete $amplicons->{$amplicon_name};
                next AMPLICON;
            }
        }
    }

    return 1;
}

sub _get_gsc_sequence_read { # in sub to overload on test
    return GSC::Sequence::Read->get(trace_name => $_[1]);
}

#< Amplicon Reads >#
sub get_method_for_get_amplicon_name_for_read_name {
    my $self = shift;

    return sprintf(
        '_get_amplicon_name_for_%s_%s_read_name', 
        $self->sequencing_center,
        $self->sequencing_platform,
    );
}

sub _get_amplicon_name_for_gsc_sanger_read_name {
    my ($self, $read_name) = @_;

    $read_name =~ /^(.+)\.[bg]\d+$/
        or return;

    return $1;
}

sub _get_amplicon_name_for_broad_sanger_read_name {
    my ($self, $read_name) = @_;

    $read_name =~ s#\.T\d+$##;
    $read_name =~ s#[FR](\w\d\d?)$#\_$1#; # or next;

    return $read_name;
}

sub get_all_amplicons_reads_for_read_name {
    my ($self, $read_name) = @_;

    unless ( $read_name ) {
        $self->error_message("No read name given to get all reads for amplicon.");
        return;
    }

    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $amplicon_name = $self->$amp_method($read_name);
    unless ( $amplicon_name ) {
        $self->error_message("Can't get amplicon name for read name ($read_name)");
        return;
    }

    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;
    my @read_names = $self->$reads_method($amplicon_name);
    unless ( @read_names ) {
        $self->error_message("No reads found for amplicon name ($amplicon_name)");
        return;
    }


    return @read_names;
}

sub get_method_for_get_all_amplicons_reads_for_read_name {
    sprintf(
        '_get_all_reads_for_%s_%s_amplicon',
        $_[0]->sequencing_center,
        $_[0]->sequencing_platform,
    );
}

sub _get_all_reads_for_gsc_sanger_amplicon {
    my ($self, $amplicon_name) = @_;
    
    my $chromat_dir = $self->chromat_dir;
    my @read_names;
    for my $read_name ( glob("$chromat_dir/$amplicon_name.*") ) {
        $read_name =~ s#$chromat_dir/##;
        $read_name =~ s#\.gz##;
        push @read_names, $read_name;
    }

    return @read_names;
}

sub _get_all_reads_for_broad_sanger_amplicon {
    die "Not implemented\n";
}

#< Contamination Screening >#
sub contamination_dir {
    return $_[0]->directory.'/contamination';
}

sub contamination_reads_dir {
    return $_[0]->contamination_dir.'/reads';
}
    
sub amplicon_fasta_file_for_contamination_screening {
    return $_[0]->contamination_dir.'/amplicon_reads.fasta';
}

sub create_contamination_dir_structure {
    my $self = shift;

    my $contamination_dir = $self->contamination_dir;
    unless ( -d $contamination_dir ) {
        return unless Genome::Sys->create_directory($contamination_dir);
    }

    my $reads_dir = $self->contamination_reads_dir;
    unless ( -d $reads_dir ) {
        return unless Genome::Sys->create_directory($reads_dir);
    }

    return 1;
}

sub create_amplicon_fasta_file_for_contamination_screening {
    my $self = shift;

    my $amplicons = $self->get_amplicons
        or return;

    $self->create_contamination_dir_structure
        or return;
    
    my $fasta_file = $self->amplicon_fasta_file_for_contamination_screening;
    unlink $fasta_file if -e $fasta_file;
    my $fasta_writer = Bio::SeqIO->new(
        '-file' => '>'.$fasta_file,
        '-fomat' => 'fasta',
    );
    for my $amplicon ( @$amplicons ) {
        for my $bioseq ( $amplicon->get_bioseqs_for_processed_reads ) {
            next unless $bioseq->length >= 11;
            $bioseq->seq( uc $bioseq->seq );
            $fasta_writer->write_seq($bioseq);
        }
    }

    return $fasta_file;
}

sub remove_contaminated_amplicons_by_reads_in_file {
    my ($self, $file) = @_;

    $self->create_contamination_dir_structure
        or return;

    my $fh = Genome::Sys->open_file_for_reading($file)
        or return;

    my %amplicons_seen;
    my $chromat_dir = $self->chromat_dir;
    my $contamination_reads_dir = $self->contamination_reads_dir;
    my $amp_method = $self->get_method_for_get_amplicon_name_for_read_name;
    my $reads_method = $self->get_method_for_get_all_amplicons_reads_for_read_name;

    while ( my $read_name = $fh->getline ) {
        chomp $read_name;
        my $amplicon_name = $self->$amp_method($read_name);
        unless ( $amplicon_name ) {
            $self->error_message("Can't get amplicon name for read name ($read_name)");
            return;
        }
        my @read_names = $self->$reads_method($amplicon_name);
        unless ( @read_names ) {
            $self->error_message("Can't get reads for amplicon name ($amplicon_name)");
            return;
        }
        # Move chromats - this effectively removes the amplicon
        # TODO create the amplicon interface here, to support sanger and 454
        #  maybe move this logic to amplicon?
        for my $read_name ( @read_names ) {
            my $from = "$chromat_dir/$read_name.gz"; 
            my $to = "$contamination_reads_dir/$read_name.gz";
            unless ( -e $from ) { # ok, i guess
                $self->error_message("Can't find trace $read_name");
                next;
            }
            unless ( File::Copy::move($from, $to) ) { # not ok
                $self->error_message("Can't move $from to $to\: $!");
                return;
            }
        }
        # TODO move phd_dir edit_dir files?
    }

    return 1;
}

1;

=pod

=head1 Name

Genome::AmpliconAssembly

=head1 Synopsis

=head1 Usage

=head1 Methods

=head2 valid_sequencing_centers

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 default_sequencing_center

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 consed_directory

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_directory_structure

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 edit_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 phd_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 chromat_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_types

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_bioseq_method_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 fasta_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 qual_file_for_type

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 get_amplicons

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 contamination_reads_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 amplicon_fasta_file_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_contamination_dir

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 create_amplicon_fasta_files_for_contamination_screening

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head2 read_is_contaminated

=over

=item I<Synopsis>

=item I<Arguments>

=item I<Returns>

=back

=head1 See Also

=head1 Disclaimer

Copyright (C) 2005 - 2009 Genome Center at Washington University in St. Louis

This module is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY or the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

=head1 Author(s)

B<Eddie Belter> I<ebelter@genome.wustl.edu>

=cut

#$HeadURL$
#$Id$
