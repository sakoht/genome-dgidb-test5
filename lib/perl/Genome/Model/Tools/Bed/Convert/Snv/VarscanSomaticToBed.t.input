1	14907	A	G	1	2	66.67%	R	3	2	40%	R	Germline	0.03846153846153868	0.9285714285714279	1	2	1	1
1	14930	A	G	2	3	60%	R	4	3	42.86%	R	Germline	0.0068649885583524735	0.8787878787878745	2	2	1	2
1	54716	C	T	1	4	80%	T	2	4	66.67%	Y	Germline	5.159958720330254E-4	0.87878787878788	2	0	2	2
1	54753	T	G	10	0	0%	T	6	3	33.33%	K	Somatic	1.0	0.0866873065015481	4	2	2	1
1	55299	C	T	2	3	60%	Y	3	2	40%	Y	Germline	0.016253869969040224	0.8968253968253947	1	2	1	1
1	69270	A	G	1	40	97.56%	G	0	50	100%	G	Germline	2.541094430744263E-52	0.45054945054945994	0	0	50	0
1	69511	A	G	0	93	100%	G	0	85	100%	G	Germline	1.612172329239279E-106	1.0	0	0	28	57
1	82609	C	G	1	2	66.67%	S	3	4	57.14%	S	Germline	0.0054179566563467545	0.8333333333333324	1	2	2	2
1	83964	A	G	6	0	0%	A	6	2	25%	R	Somatic	1.0	0.30769230769230677	3	3	1	1
1	84244	A	C	2	2	50%	M	6	2	25%	M	Germline	0.04658385093167752	0.9333333333333328	2	4	1	1
1	86065	G	C	4	3	42.86%	S	4	3	42.86%	S	Germline	0.007971014492753606	0.7039627039627032	2	2	2	1
1	87409	C	T	3	0	0%	C	0	4	100%	T	Somatic	1.0	0.02857142857142854	0	0	0	4
1	109534	A	C	7	2	22.22%	M	1	2	66.67%	M	Germline	0.04658385093167752	0.2363636363636356	1	0	2	0
1	118617	T	C	2	3	60%	Y	3	2	40%	Y	Germline	0.016253869969040224	0.8968253968253947	2	1	1	1
1	129285	G	A	1	5	83.33%	A	3	6	66.67%	R	Germline	2.4987506246876623E-5	0.9076923076923035	0	3	0	6
1	245024	C	A	4	5	55.56%	M	5	6	54.55%	M	Germline	7.265329845975073E-5	0.6890450107168408	1	4	3	3
1	569492	T	C	4	23	85.19%	C	1	3	75%	C	Germline	8.099893299651852E-13	0.8882885746067052	0	1	0	3
1	589082	G	A	0	3	100%	A	1	2	66.67%	R	Germline	0.007575757575757551	0.49999999999999994	0	1	0	2
1	736289	T	A	1	2	66.67%	W	1	2	66.67%	W	Germline	0.030303030303030318	0.7999999999999995	1	0	1	1
1	745347	T	C	11	0	0%	T	15	3	16.67%	Y	Somatic	1.0	0.22331691297208245	15	0	3	0
1	753541	G	A	1	4	80%	A	0	4	100%	A	Germline	2.0567667626491322E-4	0.5555555555555558	0	0	0	4
1	754964	C	T	0	3	100%	T	6	3	33.33%	Y	Germline	0.0068649885583524735	0.09090909090909075	4	2	1	2
1	765887	A	C	9	0	0%	A	3	2	40%	M	Somatic	1.0	0.10989010989011024	2	1	1	1
1	769717	A	G	9	0	0%	A	4	2	33.33%	R	Somatic	1.0	0.1428571428571436	1	3	1	1
1	769829	C	A	0	7	100%	A	0	6	100%	A	Germline	9.614829913658729E-8	1.0	0	0	0	6
1	770568	A	G	0	7	100%	G	0	4	100%	G	Germline	1.4175710770137988E-6	1.0	0	0	1	3
1	774736	A	C	0	6	100%	C	0	4	100%	C	Germline	5.412544112234578E-6	1.0	0	0	2	2
1	774923	G	C	0	4	100%	C	0	5	100%	C	Germline	2.056766762649123E-5	1.0	0	0	2	3
1	798026	C	T	0	3	100%	T	0	3	100%	T	Germline	0.0010822510822510842	1.0	0	0	1	2
1	798959	G	A	3	2	40%	R	2	2	50%	R	Germline	0.041176470588235946	0.6428571428571408	0	2	1	1
1	799463	T	C	0	3	100%	C	0	11	100%	C	Germline	2.4927336813189342E-8	1.0	0	0	6	5
1	801467	G	C	3	3	50%	S	3	2	40%	S	Germline	0.017543859649122945	0.8246753246753257	3	0	1	1
1	801943	C	T	0	4	100%	T	0	3	100%	T	Germline	2.9137529137529E-4	1.0	0	0	2	1
1	808922	G	A	0	60	100%	A	0	13	100%	A	Germline	1.7005940239329643E-43	1.0	0	0	7	6
1	808928	C	T	52	13	20%	Y	5	9	64.29%	Y	Germline	1.0	0.0018966685629848313	2	3	5	4
1	814495	C	A	6	2	25%	M	1	2	66.67%	M	Germline	0.045112781954887035	0.2787878787878776	1	0	2	0
1	814535	A	G	4	3	42.86%	R	2	2	50%	R	Germline	0.017543859649122945	0.6515151515151519	2	0	2	0
1	814560	G	T	4	2	33.33%	K	1	2	66.67%	K	Germline	0.041176470588235946	0.4047619047619044	1	0	2	0
1	814609	A	T	3	2	40%	W	1	2	66.67%	W	Germline	0.03846153846153868	0.49999999999999967	0	1	0	2
1	814611	A	C	3	2	40%	M	1	2	66.67%	M	Germline	0.03846153846153868	0.49999999999999967	0	1	0	2
1	820944	G	T	3	0	0%	G	0	3	100%	T	Somatic	1.0	0.04999999999999994	0	0	1	2
1	820957	C	T	3	0	0%	C	0	3	100%	T	Somatic	1.0	0.04999999999999994	0	0	1	2
1	821729	A	C	2	3	60%	M	1	3	75%	C	Germline	0.004524886877828076	0.5952380952380945	0	1	0	3
1	821772	G	C	1	4	80%	C	1	4	80%	C	Germline	3.5722791140748014E-4	0.7777777777777751	0	1	1	3
1	821787	C	T	1	4	80%	T	1	4	80%	T	Germline	3.5722791140748014E-4	0.7777777777777751	0	1	1	3
1	821793	G	A	1	3	75%	A	1	4	80%	A	Germline	0.0011312217194570148	0.72222222222222	0	1	1	3
1	823504	T	C/G	0	1	100%	C	4	2	33.33%	K	Unknown	1.0	0.42857142857142827	3	1	1	1
1	823505	T	C	3	0	0%	T	3	4	57.14%	Y	Somatic	1.0	0.16666666666666616	1	2	3	1
1	824094	G	A	1	4	80%	A	3	2	40%	R	Germline	0.0054179566563467545	0.9761904761904733	0	3	0	2
1	824904	G	C	2	3	60%	S	0	3	100%	C	Germline	0.00349650349650351	0.35714285714285743	0	0	2	1
