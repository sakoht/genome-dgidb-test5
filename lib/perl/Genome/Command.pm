package Genome::Command;

use strict;
use warnings;

use Genome;

class Genome::Command {
    is => 'Command::Tree',
};
         
# This map allows the top-level genome commands to be whatever
# we wish, instead of having to match the directory structure.
my %command_map = (
    'disk' => 'Genome::Disk::Command',
    'feature-list' => 'Genome::FeatureList::Command',
    'gene-name' => 'Genome::GeneName::Command',
    'individual' => 'Genome::Individual::Command',
    'instrument-data' => 'Genome::InstrumentData::Command',
    'library' => 'Genome::Library::Command',
    'model' => 'Genome::Model::Command',
    'model-group' => 'Genome::ModelGroup::Command',
    'notable' => 'Genome::Notable::Command',
    'population-group' => 'Genome::PopulationGroup::Command',
    'processing-profile' => 'Genome::ProcessingProfile::Command',
    'project' => 'Genome::Project::Command',
    'project-part' => 'Genome::ProjectPart::Command',
    'report' => 'Genome::Report::Command',
    'sample' => 'Genome::Sample::Command',
    'subject' => 'Genome::Subject::Command',
    'sys' => 'Genome::Sys::Command',
    'task' => 'Genome::Task::Command',
    'taxon' => 'Genome::Taxon::Command',
    'tools' => 'Genome::Model::Tools',
    'db' => 'Genome::Db',
);

for my $class (values %command_map) {
    eval "use $class";
    die $@ if $@;
}

$Genome::Command::SUB_COMMAND_MAPPING = \%command_map;

1;
