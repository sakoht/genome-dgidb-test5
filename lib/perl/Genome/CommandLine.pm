package Genome::CommandLine;

use strict;
use warnings;

use Genome;

class Genome::CommandLine {
    is => 'Command::Tree',
};

sub command_tree_source_classes {
    return ('Genome::Model::Tools','Genome::Command');
}

sub _categorize_sub_commands {
    my $class = shift;

    my @sub_command_classes = $class->sorted_sub_command_classes;
    my %categories;
    my @order;
    for my $sub_command_class (@sub_command_classes) {
        next if $sub_command_class->_is_hidden_in_docs();
        #next unless $sub_command_class =~ /Genome::B/;

        my $category;
        if ($sub_command_class =~ /Genome::Model::Tools/) {
            $category = 'tools';
        }
        else {
            $category = 'system';
        }

        unless (exists $categories{$category}) {
            if ($category) {
                push(@order, $category)
            } else {
                unshift(@order, '');
            }
            $categories{$category} = [];
        }
        push(@{$categories{$category}}, $sub_command_class);
    }

    return (\@order, \%categories);
}

1;
