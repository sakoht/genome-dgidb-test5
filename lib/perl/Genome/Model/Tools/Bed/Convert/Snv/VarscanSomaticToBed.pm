package Genome::Model::Tools::Bed::Convert::Snv::VarscanSomaticToBed;

use strict;
use warnings;

use Genome;

class Genome::Model::Tools::Bed::Convert::Snv::VarscanSomaticToBed {
    is => ['Genome::Model::Tools::Bed::Convert::Snv'],
};

sub help_brief {
    "Tools to convert varscan-somatic SNV format to BED.",
}

sub help_synopsis {
    my $self = shift;
    return <<"EOS"
  gmt bed convert snv varscan-somatic-to-bed --source snps_all_sequences --output snps_all_sequences.bed
EOS
}

sub help_detail {                           
    return <<EOS
    This is a small tool to take SNV calls in var-scan format and convert them to a common BED format (using the first four columns + quality).
EOS
}

sub process_source {
    my $self = shift;
    
    my $input_fh = $self->_input_fh;
    
    while(my $line = $input_fh->getline) {
        my ($chromosome, $position, $reference, undef,$depth1, $depth2, undef, undef, undef,$qual , undef,$consensus, @extra) = split("\t", $line);
        $DB::single=1; 
        no warnings qw(numeric);
        next unless $position eq int($position); #Skip header line(s)
        use warnings qw(numeric);
        
        #position => 1-based position of the SNV
        #BED uses 0-based position of and after the event
        my $depth = $depth1+$depth2;
        $self->write_bed_line($chromosome, $position - 1, $position, $reference, $consensus, $qual, $depth);
    }
    
    return 1;
}

1;
