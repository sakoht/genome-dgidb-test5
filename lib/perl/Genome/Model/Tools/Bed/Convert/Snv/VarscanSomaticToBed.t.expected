1	137	138	T/Y	30	170
1	143	144	T/Y	31	135
1	179	180	T/Y	31	329
1	230	231	C/M	33	132
1	233	234	C/Y	33	153
1	247	248	A/W	34	69
1	249	250	A/M	31	70
1	254	255	A/W	33	42
1	256	257	A/M	33	60
1	290	291	C/Y	27	17
1	314	315	C/Y	28	19
1	320	321	C/Y	29	26
1	326	327	T/Y	28	27
1	327	328	A/W	27	32
1	389	390	C/M	32	110
1	392	393	C/Y	34	107
1	396	397	C/M	32	104
1	439	440	C/M	34	65
1	442	443	C/Y	34	88
1	469	470	G/R	32	27
1	471	472	G/S	32	30
1	472	473	G/S	33	29
1	481	482	G/S	32	31
1	518	519	G/S	35	22
1	582	583	G/R	33	17
1	602	603	G/R	34	16
1	4769	4770	A/R	35	49
1	4792	4793	A/G	34	56
1	4980	4981	A/R	35	16
1	5052	5053	G/R	33	14
1	5073	5074	T/G	0	15
